src/operator/numpy/linalg/np_gesvd-inl.h

```diff
@@ -0,0 +1,302 @@
+      Y[offy + i] = X[offx + i * ldx + i];
+      for (int j = 0; j < m; ++j) {
+        numer = L[offl + j];
+        X[offx + i * ldx + j] *= numer;
+      }
+    }
+  }
+};
+
+struct GesvdBackHelper_G2 {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int k, int m, int n, DType* X, int ldx,
+                                  DType* L, int ldl, DType* dL, int lddl,
+                                  DType* Y, int ldy) {
+    const int offx(k * m * ldx);
+    const int offl(k * ldl);
+    const int offdl(k * lddl);
+    const int offy(k * ldy);
+    const DType eps(gesvd_back_helper_eps(X));
+    DType denom1(0.0), denom2(0.0), elem(0.0);
+
+    for (int i = 0; i < m; ++i) {
+      for (int j = i + 1; j < m; ++j) {
+        denom1 = L[offl + i] - L[offl + j];
+        denom2 = L[offl + i] + L[offl + j];
+        if (denom1 < eps) denom1 = eps;
+        if (denom2 < eps) denom2 = eps;
+        elem = (X[offx + i * ldx + j] - X[offx + j * ldx + i]) / denom1 / denom2;
+        X[offx + i * ldx + j] = elem * L[offl + j];
+        X[offx + j * ldx + i] = elem * L[offl + i];
+      }
+      X[offx + i * ldx + i] = -Y[offy + i] + dL[offdl + i];
+    }
+  }
+};
+
+struct gesvd_backward {
+  template<typename xpu, typename DType>
+  static void op(const Tensor<xpu, 3, DType>& dUT,
+                 const Tensor<xpu, 2, DType>& dL,
+                 const Tensor<xpu, 3, DType>& dV,
+                 const Tensor<xpu, 3, DType>& UT,
+                 const Tensor<xpu, 2, DType>& L,
+                 const Tensor<xpu, 3, DType>& V,
+                 const Tensor<xpu, 3, DType>& dA,
+                 const Tensor<xpu, 3, DType>& tempM,
+                 const Tensor<xpu, 2, DType>& tempMd,
+                 Stream<xpu>* s, const nnvm::NodeAttrs& attrs) {
+    // Backward of (UT, L, V) = gesvd(A)
+    using namespace mxnet_op;
+    if (dA.dptr_ != dV.dptr_) {
+      Copy(dA, dV, s);
+    }
+    // From here on, we work on dA only
+    int k = dA.size(0), m = dA.size(1), n = dA.size(2);
+
+    // Need temporal space, same shape as dUT
+    // invdV:
+    Kernel<GesvdBackHelper_dV, xpu>::Launch
+      (s, k, m, n, L.dptr_, L.stride_, dA.dptr_, dA.stride_);
+
+    // G1:
+    // This is just to make sure there are no invalid values (NaN, infinity) in tempM and tempMd
+    tempM.FlatTo1D() = 0;
+    tempMd.FlatTo1D() = 0;
+    gemm::op(dA, V, tempM, DType(1.0), DType(0.0), false, true, s);
+    Kernel<GesvdBackHelper_G1, xpu>::Launch
+      (s, k, m, n, tempM.dptr_, tempM.stride_,
+       L.dptr_, L.stride_, tempMd.dptr_, tempMd.stride_);
+    gemm::op(dUT, UT, tempM, DType(1.0), DType(1.0), true, false, s);
+
+    // G2:
+    Kernel<GesvdBackHelper_G2, xpu>::Launch
+      (s, k, m, n, tempM.dptr_, tempM.stride_,
+       L.dptr_, L.stride_, dL.dptr_, dL.stride_,
+       tempMd.dptr_, tempMd.stride_);
+
+    // G3:
+    gemm::op(tempM, V, dA, DType(1.0), DType(1.0), false, false, s);
+    for (int i = 0; i < n; i += m) {
```

Comment:
Here I go through the blocks (0:m, i: i + m). Other attrs of tensors like stride_, stream_ are not changed, as you said. The ncols is the m2 you just mentioned. It is used for the last block.